engine/schema/src/com/cloud/upgrade/dao/Upgrade481to490.java

```diff
@@ -53,6 +62,139 @@ public boolean supportsRollingUpgrade() {
+                "ADD KEY `fk_account__role_id` (`role_id`), " +
+                "ADD CONSTRAINT `fk_account__role_id` FOREIGN KEY (`role_id`) REFERENCES `roles` (`id`);";
+        try (PreparedStatement pstmt = conn.prepareStatement(alterTableSql)) {
+            pstmt.executeUpdate();
+            s_logger.info("Altered cloud.account table and added column role_id");
+        } catch (SQLException e) {
+            if (e.getMessage().contains("role_id")) {
+                s_logger.warn("cloud.account table already has the role_id column, skipping altering table and migration of accounts");
+                return;
+            } else {
+                throw new CloudRuntimeException("Unable to create column quota_calculated in table cloud_usage.cloud_usage", e);
+            }
+        }
+        migrateAccountsToDefaultRoles(conn, roleTypes);
+    }
+
+    private void migrateAccountsToDefaultRoles(final Connection conn, final RoleType[] roleTypes) {
+        try (PreparedStatement selectStatement = conn.prepareStatement("SELECT `id`, `type` FROM `cloud`.`account`;");
+             ResultSet selectResultSet = selectStatement.executeQuery()) {
+            while (selectResultSet.next()) {
+                Long accountId = selectResultSet.getLong(1);
+                Short accountType = selectResultSet.getShort(2);
+                Long roleId = null;
+                for (RoleType roleType : roleTypes) {
+                    if (roleType.getAccountType() == accountType) {
+                        roleId = roleType.getId();
+                        break;
+                    }
+                }
+                if (roleId == null) {
+                    continue;
+                }
+                try (PreparedStatement updateStatement = conn.prepareStatement("UPDATE `cloud`.`account` SET role_id = ? WHERE id = ?;")) {
+                    updateStatement.setLong(1, roleId);
+                    updateStatement.setLong(2, accountId);
+                    updateStatement.executeUpdate();
+                    updateStatement.close();
+
+                } catch (SQLException e) {
+                    s_logger.error("Failed to update cloud.account role_id for account id:" + accountId + " with exception: " + e.getMessage());
+                    throw new CloudRuntimeException("Exception while updating cloud.account role_id", e);
+                }
+            }
+        } catch (SQLException e) {
+            throw new CloudRuntimeException("Exception while migrating existing account table's role_id column to a role based on account type", e);
+        }
+        s_logger.debug("Done migrating existing accounts to use one of default roles based on account type");
+    }
+
+    private void setupRolesAndPermissionsForDynamicRBAC(final Connection conn) {
+        try ( PreparedStatement selectStatement = conn.prepareStatement("SELECT * FROM `cloud`.`roles`");
+              ResultSet resultSet = selectStatement.executeQuery()) {
```

Comment:
no need, multiple resources should be automatically closed with try-with-resource syntax